src/stats/mod.rs

```diff
@@ -0,0 +1,208 @@
+// TODO: Everything in this file
+#![allow(dead_code)]
+#![allow(unused_variables)]
+
+use std::{
+    net::IpAddr,
+    sync::{
+        LazyLock,
+        atomic::{AtomicU64, Ordering},
+    },
+};
+
+use chrono::{DateTime, Duration, FixedOffset, NaiveDate, Utc};
+use log::info;
+use papaya::HashMap;
+use sarlacc::Intern;
+
+const IP_TRACKING_TTL: chrono::TimeDelta = Duration::days(1);
+pub const TIMEZONE: chrono::FixedOffset = FixedOffset::west_opt(5 * 3600).unwrap();
+
+pub static UNKNOWN_ORIGIN: LazyLock<Intern<str>> = LazyLock::new(|| Intern::from_ref("unknown"));
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+struct IpInfo {
+    last_seen: DateTime<Utc>,
+    started_from: Intern<str>,
+}
+
+#[derive(Debug, Default)]
+struct AggregatedStats {
+    // (Date (with timezone `TIMEZONE`), From, To, Started From) → Count
+    #[expect(clippy::type_complexity)]
+    counters: HashMap<(NaiveDate, Intern<str>, Intern<str>, Intern<str>), AtomicU64>,
+}
+
+#[derive(Debug, Default)]
+pub struct Stats {
+    aggregated: AggregatedStats,
+    ip_tracking: HashMap<IpAddr, IpInfo>,
+}
+
+impl Stats {
+    pub fn new() -> Stats {
+        Stats {
+            aggregated: AggregatedStats::default(),
+            ip_tracking: HashMap::new(),
+        }
+    }
+
+    pub fn redirected(&self, ip: IpAddr, from: Intern<str>, to: Intern<str>) {
+        self.redirected_impl(ip, from, to, Utc::now());
+    }
+
+    fn redirected_impl(&self, ip: IpAddr, from: Intern<str>, to: Intern<str>, now: DateTime<Utc>) {
+        let ip_info = *self.ip_tracking.pin().update_or_insert(
+            ip,
+            |ip_info| {
+                let mut info = *ip_info;
+                info.last_seen = now;
+                info
+            },
+            IpInfo {
+                last_seen: now,
+                started_from: from,
+            },
+        );
+
+        let date = now.with_timezone(&TIMEZONE).date_naive();
+
+        let pinned_map = self.aggregated.counters.pin();
+        let counter =
+            pinned_map.get_or_insert((date, from, to, ip_info.started_from), AtomicU64::new(0));
+
+        counter.fetch_add(1, Ordering::Relaxed);
+    }
+
+    pub fn prune_seen_ips(&self) {
+        self.prune_seen_ips_impl(Utc::now());
+    }
+
+    #[expect(clippy::cast_possible_wrap)]
+    fn prune_seen_ips_impl(&self, now: DateTime<Utc>) {
```

Comment:
Do we need a separate function? It seems like this could just be moved into `prune_seen_ips()` and the parameter can become a variable.

Reply:
`prune_seen_ips_impl` (and the other _impl things) are used in tests to set the timestamp exactly, which is why they're separate functions.